### PR TITLE
Build provider cluster reconcilers at program init

### PR DESCRIPTION
**Pull request by g-gaston**
*Description of changes:*

Before this, provider cluster reconcilers were built dynamically on each
reconciliation loop. This required injecting all secondary dependencies
in the main cluster reconciler so they could be passed down to the
provider reconcilers when needed. This doesn't really scale with multiple
providers and made development difficult.

Instead, now we build the necessary providers during program
initialization. We check which are the CAPI providers that are installed
and we build the appropriate provider reconcilers based on that.

I made some minor refactors that are somehow not related with this
improvement but made working with this code easier and in a way shorten
the length of this PR. This includes: getting rid of the "abstract"
provider cluster reconciler and drying out some tests.

*Testing (if applicable):*

Unit tests

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: controllers/cluster_controller.go [hunk 1]
```diff
@@ -40,33 +33,21 @@ const (
...
-// TODO: this is not ideal and will need a refactor. I will follow up but for now this
-// allows us to decouple the cluster reconciler main logic from provider specific logic
-type ProviderReconcilerBuilder func(datacenterKind string, client client.Client, log logr.Logger, validator *vsphere.Validator, defaulter *vsphere.Defaulter, tracker *remote.ClusterCacheTracker) (ProviderClusterReconciler, error)
-
-func NewClusterReconciler(client client.Client, log logr.Logger, scheme *runtime.Scheme, govc *executables.Govc, tracker *remote.ClusterCacheTracker, buildProviderReconciler ProviderReconcilerBuilder) *ClusterReconciler {
-	validator := vsphere.NewValidator(govc, &networkutils.DefaultNetClient{})
-	defaulter := vsphere.NewDefaulter(govc)
-
+// TODO: rename
+func NewNewClusterReconciler(client client.Client, log logr.Logger, registry ProviderClusterReconcilerRegistry) *ClusterReconciler {
```
Comment (vivek-koppuru) [inline controllers/cluster_controller.go:46]: This is to rename at a later PR?

File: controllers/cluster_controller_test_test.go [hunk 2]
```diff
@@ -67,7 +65,7 @@ func TestClusterReconcilerEnsureOwnerReferences(t *testing.T) {
 	cb := fake.NewClientBuilder()
 	cl := cb.WithRuntimeObjects(objs...).Build()
 
-	r := controllers.NewClusterReconciler(cl, nullLog(), cl.Scheme(), nil, nil, newDummyProviderReconcilerBuilder())
+	r := controllers.NewNewClusterReconciler(cl, nullLog(), newRegistryForDummyProviderReconciler())
```
Comment (vivek-koppuru) [inline controllers/cluster_controller_test_test.go:68]: What is the purpose of the `_test_test` file?

File: pkg/controller/clusters/registry.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+	anywherev1 "github.com/aws/eks-anywhere/pkg/api/v1alpha1"
+	"github.com/aws/eks-anywhere/pkg/controller"
+)
+
+type ProviderClusterReconciler interface {
+	Reconcile(ctx context.Context, cluster *anywherev1.Cluster) (controller.Result, error)
+}
+
+type ProviderClusterReconcilerRegistry struct {
```
Comment (vivek-koppuru) [inline pkg/controller/clusters/registry.go:14]: What is a registry here?

File: controllers/cluster_controller_test_test.go (hunk 2 above)
Comment (g-gaston) [reply controllers/cluster_controller_test_test.go:68]: The original `_test` file is in the `controllers` package as opposed to `controllers_test`

This file was added to be able to add tests under `controllers_test` until the old tests get moved to the proper test package

File: pkg/controller/clusters/registry.go (hunk 3 above)
Comment (g-gaston) [reply pkg/controller/clusters/registry.go:14]: I'll add comments

File: controllers/cluster_controller.go (hunk 1 above)
Comment (g-gaston) [reply controllers/cluster_controller.go]: Ooops, thanks, fixed

File: pkg/controller/clusters/registry.go (hunk 3 above)
Comment (g-gaston) [reply pkg/controller/clusters/registry.go:16]: Done, let me know

**Review by vivek-koppuru — APPROVED** [review]

**Pull request merged**